Make the max default mempool limiter be the same as the default min relay fee

ptschip opened the pull request:
This will keep us in line , by default, with the rest of the Cash network and will prevent our mempools from getting out of sync, if/when there are large and sudden spikes in txn activity, and help with Graphene as well as 0 conf.  But still an operator can modify these setting if they wish.

ptschip committed:
Move variable initializations for the mempool limiter to main.h

This is a move only, no other changes.

ptschip committed:
Make the DEFAULT_MAXLIMITERTXFEE equal to the DEFAULT_MIN_RELAY_TX_FEE

This way we will not trample on zero-conf since our limiter will by
default not rise higher than the min relay fee defaults. However
operators can still manually change their mempool limiter settings to
whatever policy they wish to enforce.

ptschip commented:
I took these commits out of 1181 as they really should be in a separate PR.

gandrewstone commented on src/main.cpp:
```diff
@@ -775,8 +775,9 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
+        static std::string default_maxlimitertxfee = std::to_string((double)DEFAULT_MAXLIMITERTXFEE / 1000);
         static const double maxFeeCutoff =
-            boost::lexical_cast<double>(GetArg("-maxlimitertxfee", DEFAULT_MAXLIMITERTXFEE));
+            boost::lexical_cast<double>(GetArg("-maxlimitertxfee", default_maxlimitertxfee.c_str()));
```
accepttomemorypoolworker is performance sensitive code.  Although these unnecessary conversion and GetArg lookups probably won't cause performance issues, simpler is also better, and this is deploying a lot of machinery (heap allocations, map lookups, string conversions) that could happen only once.  

I propose we make maxlimitertxfee and minlimitertxfee a CTweak (it does make sense to be able to change them at runtime, right?).  Although CTweaks typically use a dotted argname notation, for these existing args you'd use the existing names.

ptschip replied:
I can make it a tweak if you want, but I also don't see any performance issue with the way it is, since it's a static and only get assigned once.

ptschip committed:
Clarify naming of maxFeeCutoff and initFeeCutoff

use the naming of dMaxLimiterTxFee and dMinLimiterTxFee to denote
these variables. This make the code easier to read.

ptschip committed:
Make both minlimitertxfee and maxlimtertxfee tweaks.

ptschip committed:
for clarity change feeCutoff to nMinRelay

ptschip committed:
Add critcal section cs_limiter

Although we don't need this now because cs_main locks this section
we will in the future need this when we reduce the scope of cs_main
in accepttomemorypoolworker.

ptschip committed:
fix format

ptschip committed:
Fix log output message for nMinRelay

We were not actually printing out the correct value which led
to a great deal of confusion as to why the values were drifting
when they shouldn't be.

ptschip commented:
@gandrewstone added the tweaks you wanted and cleaned up a bit.  Also added a crit sec for the memlimiter which isn't really needed now, but will be easily forgotten that we need it, when we do multi-threaded txn validation in the future.

ptschip committed:
Use Tweak version of maxlimitertxfee and minlimitertxfee only

gandrewstone commented on src/main.cpp:
```diff
@@ -775,88 +775,110 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
-        // nFreeLimit = nFreeLimit + ((double)(DEFAULT_LIMIT_FREE_RELAY - nFreeLimit) / pow(1.0 - 1.0/86400,
-        // (double)(nNow - nLastTime)));
-        nFreeLimit /= std::pow(1.0 - 1.0 / 86400, (double)(nNow - nLastTime));
+        static double _dMinLimiterTxFee = dMinLimiterTxFee.value;
```
can these now just be nonstatic locals, defined inside the lock?

ptschip replied:
it has to be static because we need it on each iteration and for each thread.  We only update it if the tweak changes.  We could use the tweak but then I want to avoid calling the tweak value for every txn for now since I think we have to look at adding a lock to the tweaks anyway.  I guess I'm prematurely avoiding having to take  the tweak lock.

gandrewstone commented on src/main.cpp:
```diff
@@ -775,88 +775,110 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
-            nLastTime = nNow;
+            uint64_t nLargestBlockSeen = LargestBlockSeen();
+
+            if (poolBytes < nLargestBlockSeen)
```
I think that this is a semantically equivalent simplification:
```
nFreeLimit /= std::pow(1.0 - 1.0 / 86400, (double)(nNow - nLastTime));
// When the mempool starts falling use an exponentially decaying ~24 hour window:
nMinRelay *= std::pow(1.0 - 1.0 / 86400, (double)(nNow - nLastTime));

if (poolBytes < nLargestBlockSeen)
{
    nMinRelay = std::max(nMinRelay, _dMinLimiterTxFee);
    nFreeLimit = std::min(nFreeLimit, (double)nLimitFreeRelay);
}
```

=>

```
if (poolBytes >= nLargestBlockSeen)
{
    nFreeLimit /= std::pow(1.0 - 1.0 / 86400, (double)(nNow - nLastTime));
    nMinRelay *= std::pow(1.0 - 1.0 / 86400, (double)(nNow - nLastTime));
}
```

ptschip replied:
Needs to remain as is.  We need to calculate nFreelimit and nMinrelay on every iteration , since we need those values if we fall through to the other if else / else statements.

gandrewstone commented on src/main.cpp:
```diff
@@ -775,88 +775,110 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
+
+                // Use an exponentially decaying ~10-minute window:
+                dFreeCount *= std::pow(1.0 - 1.0 / 600.0, (double)(nNow - nLastTime));
+                nLastTime = nNow;
```
not needed (set below)

ptschip replied:
done

gandrewstone commented on src/init.cpp:
```diff
@@ -752,6 +752,20 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
     if (nConnectTimeout <= 0)
         nConnectTimeout = DEFAULT_CONNECT_TIMEOUT;
 
+    // Set the maxlimitertxfee tweak
```
This if block isn't needed.  CTweaks are automatically loaded with the information from mapArgs

ptschip replied:
took this out.

gandrewstone commented on src/init.cpp:
```diff
@@ -760,27 +774,32 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
-        if (ParseMoney(ss.str(), n))
-            ::minRelayTxFee = CFeeRate(n);
+        if (ParseMoney(ss.str(), nAmt))
+            ::minRelayTxFee = CFeeRate(nAmt);
```
This code sets minRelayTxFee only if minlimitertxfee is explicitly set by the user.  Instead, can we set minRelayTxFee based the value minlimitertxfee no matter what --presumably the default minlimitertxfee results in the expected default minRelayTxFee.  Otherwise we get into a situation where if minlimitertxfee is set to the default value, the behavior is different than if minlimitertxfee not touched.  If we can do this, the "if (mapArgs.count("-minlimitertxfee"))" can be removed

ptschip replied:
all taken out, not needed anymore

gandrewstone commented on src/init.cpp:
```diff
@@ -760,27 +774,32 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
         ss << minrelaytxfee;
-        if (ParseMoney(ss.str(), n))
-            ::minRelayTxFee = CFeeRate(n);
+        if (ParseMoney(ss.str(), nAmt))
```
We should not need financial number parsing (ParseMoney) here.  If we do, we have a bug because we've already read the configuration value in as a double and divide by 100000 so imprecision has occurred.  So instead of this code, we should be able to do nAmt = dMinLimiterTxFee.value * (BTC / 100000)

ptschip replied:
ok i took this all out. not needed anymore.

ptschip committed:
Always set minRelayTxFee in init.cpp

ptschip committed:
Move UnlimitedSetup() to the beginning of AppInit2

ptschip committed:
Fix fundrawtransaction.py

We need to launch the nodes with a minlimitertxfee=1.0 so that we
have a relayfee of 1000 sat per KB othewise this test will not work.

ptschip committed:
Remove redundant nLastTime assignment.
